subsys/bluetooth/mesh/adv_legacy.c

```diff
@@ -222,6 +222,13 @@ void bt_mesh_adv_init(void)
 int bt_mesh_adv_enable(void)
 {
 	k_thread_start(&adv_thread_data);
+	k_thread_resume(&adv_thread_data);
+	return 0;
+}
+
+int bt_mesh_adv_disable(void)
+{
+	k_thread_suspend(&adv_thread_data);
```

Comment:
If I'm not mistaken, this will only not let schedule the adv thread, but the adv pool will not be purged, so the behavior for legacy and extended advertiser will be different.

Reply:
That is true, is there some corresponding functionality that should be used in regards to k_threads?

Reply:
I'm not sure, but as an idea (not the only one, of course), is to add a flag which is constantly checked in the adv thread. If the flag is set, it purges the pool like adv_ext does and then exits the thread. Then `bt_mesh_adv_enable()` starts the thread again.

Reply:
Changed to checking a flag in the adv thread, then aborting the thread once the adv thread exits (using a semaphore to ensure the tread is not in the middle of some work). 
`bt_mesh_adv_enable()` runs the thread, but requires `bt_mesh_adv_init()` to be called beforehand to create the thread again after abortion. Changed the `bt_mesh_resume()` function to call `bt_mesh_adv_init()` if legacy advertising is used.

Reply:
Hm, if I'm not mistaken, the adv buffers that were pushed to the adv pool but not sent yet, will stay in the pool until the thread is started again. This creates different behavior from the extended advertiser, where all buffers will be dropped and the upper layers will be notified over `bt_mesh_send_cb.start` callback immediately. Probably, we should keep the thread alive, but if the flag is set, for every adv buffer that falls into the adv pool, call `bt_mesh_adv_send_start()` with `-ENODEV` error (like for the extended advertiser if I'm not mistaken)? Between this code:
https://github.com/zephyrproject-rtos/zephyr/blob/8b58048d02015e21a8c5b71857fa4289cd1cb8da/subsys/bluetooth/mesh/adv_legacy.c#L181-L184
Something like:
```
if (!enabled) {
        bt_mesh_adv_send_start(0, -ENODEV, BT_MESH_ADV(buf));
        net_buf_unref(buf);
        continue;
}
```

The other option is to prevent allocating any adv buffer in `bt_mesh_adv_create` if the stack is suspended. But we need to check if return value of `bt_mesh_adv_create` is correctly handled in the stack.

Reply:
Keeping the thread alive works when just suspending mesh, but not when disabling Bluetooth. Not joining/aborting the thread before disabling BT results in an error:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/91412738/75c1687e-f6c0-4ab7-b3be-97c8970bfa46)

I do not immediately see the problem in terminating the thread, by suspending mesh we will not send anything until we resume mesh? Is there some cleanup we can instead do when the adv thread exits its while loop?

Reply:
The problem in the different behavior of legacy and ext advertiser. As I suggested, the other option is that prevent allocating adv buffers in `bt_mesh_adv_create` if mesh is suspended. The adv buffers that are already in the pool should be released and the start callback should be called with error code.

Reply:
I believe this is already implemented:
https://github.com/zephyrproject-rtos/zephyr/blob/934a09550dd17406426ac92aac89761cbe506520/subsys/bluetooth/mesh/adv.c#L114-L117

For releasing the adv buffers already in the pool, is it sufficient to call `bt_mesh_adv_send_start(0, -ENODEV, BT_MESH_ADV(buf));`, or is there something I am missing?

Reply:
I believe this is already implemented:
https://github.com/zephyrproject-rtos/zephyr/blob/934a09550dd17406426ac92aac89761cbe506520/subsys/bluetooth/mesh/adv.c#L114-L117

For releasing the adv buffers already in the pool, is it sufficient to call `bt_mesh_adv_send_start(0, -ENODEV, BT_MESH_ADV(buf));`, or is there something I am missing?

Reply:
Discussed offline: Updated to empty the adv buffers by using `bt_mesh_adv_buf_get` to empty the pool upon exiting the adv thread.